core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -598,20 +649,38 @@ public TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec update
       builder.add(freshSpec);
     }
 
+    // determine the next order id
+    int maxOrderId = sortOrders.stream().mapToInt(SortOrder::orderId).max().orElse(INITIAL_SORT_ORDER_ID);
+    int nextOrderId = maxOrderId + 1;
```

Comment:
This block of changes 652 -668 Is confusing to me, could add some more comments? I'm mostly confused about when we add "freshSortOrder" , it looks like we ignore it entirely if an equivalent SortOrder is there?

Reply:
I've matched the logic we have for `PartitionSpec`. If we have an equivalent partition spec, we use it. I'll add a comment.

Reply:
This is used when we are running a `REPLACE TABLE` operation. We want to keep the history, but the schema and partition spec can be completely new. When the schema and partition spec are new, we assign fresh IDs to the schema and rebuild the partition spec (and sort order) to use the newly assigned IDs. That's why we use the `freshSpec` and `freshSortOrder` methods here as well as when creating new table metadata -- because we don't trust the IDs that were assigned in the schema that was passed in.

I should note that when we update a schema, we also rebuild the specs and sort orders. But in those cases, the IDs aren't what we are changing. What may change are the field names in the schema.